### PR TITLE
tarantoolctl: fix default host for connect

### LINKED ISSUE #7479: tarantoolctl cannot connect to tarantool 2.10 without specifying host
**Steps to reproduce**
tarantool-2.10.0-1.el8
`tarantoolctl connect 3302`

**Actual behavior**
Connection is not established if you do not specify host `tarantoolctl connect 127.0.0.1:3302`

**Expected behavior**
Connection is successful

### PR DESCRIPTION
**Pull request by GRISHNOV**
Made host=localhost by default for tarantoolctl connect. Now, when specifying only the port, the connection is made to localhost.

Fixes #7479

### CODE REVIEW

File: changelogs/unreleased/igrishnov/gh-7479-bug-fix-default-uri-tarantoolctl-connect.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+## bugfix/tarantoolctl
+
+* Made host=localhost by default for tarantoolctl connect. Now, when specifying only the port, the connection is made to localhost.
```
Comment (Totktonada) [inline changelogs/unreleased/igrishnov/gh-7479-bug-fix-default-uri-tarantoolctl-connect.md:3]: Nit: Keep it within 80 symbols wide and leave an issue number (see other entries for examples).

File: src/box/lua/console.lua [hunk 2]
```diff
@@ -818,6 +818,9 @@ local function connect(uri, opts)
     if u == nil or u.service == nil then
         error('Usage: console.connect("[login:password@][host:]port")')
     end
+    if not u.host then
+        u.host = 'localhost'
+    end
```
Comment (Totktonada) [inline src/box/lua/console.lua:823]: AFAIU, it works till some version, but degraded then. Did you look, what was the reason of the degradation?

Comment (GRISHNOV) [reply src/box/lua/console.lua:823]: At the moment, the logic of the work looks like this:

If we call `tarantoolctl connect 3301`, then after [this](https://github.com/tarantool/tarantool/blob/master/src/box/lua/console.lua#L816) we have `u = {["service"] = 3301}`.

Next , the sequence of function calls with arguments is as follows:

* net_box.connect(u.host=`nil`, u.service=`3301`, {})
* net_box.parse_connect_params(host_or_uri=`nil`, ...)

According to the logic of net_box.parse_connect_params, `host_or_uri` gets from `u.host`.

I believe that my solution that has been proposed with u.host=`localhost` before `net_box.connect` call allows to correct the situation. This default assignment goes next to the call of `urilib.parse(tostring(uri))` and checking whether the user forgot to specify the port. The location in this part of the code is expected from the point of view of code readability and [this](https://github.com/tarantool/tarantool/blob/master/src/box/lua/console.lua#L819) requirement.

Now the rationale for why  it worked before and stopped now:

AFAIU, the sequence of function calls with arguments after `urilib.parse` was as follows (everything is also being considered `tarantoolctl connect 3301`):

* net_box.establish_connection(u.host=`nil`, u.service=`3301`, opts.timeout=...)
* socket.tcp_connect(host=u.host=`nil`, post=u.service=`3301`, ...)
* socket.getaddrinfo(host=`nil`, port=`3301`, ...)

So, [socket.getaddrinfo](https://www.tarantool.io/en/doc/latest/reference/reference_lua/socket/#socket-getaddrinfo) return info about `localhost` when we call it with host=`nil`. 

Now we crash at [this](https://github.com/tarantool/tarantool/blob/master/src/box/lua/net_box.lua#L145) moment, and up to this moment the function `socket.getaddrinfo` is not called, so this bug has appeared.

File: changelogs/unreleased/igrishnov/gh-7479-bug-fix-default-uri-tarantoolctl-connect.md (hunk 1 above)
Comment (GRISHNOV) [reply changelogs/unreleased/igrishnov/gh-7479-bug-fix-default-uri-tarantoolctl-connect.md]: Fixed

File: src/box/lua/console.lua (hunk 2 above)
Comment (Totktonada) [reply src/box/lua/console.lua:823]: I was interested in a commit that introduces the regression. My testing code and `git bisect` results are below.

Run tarantool that listens on 3301 (I have `box.schema.user.grant('guest', 'super')` written to the snapshot):

```sh
$ tarantool -e 'box.cfg({listen = 3301})'
```

Run the test:

```sh
$ ./src/tarantool extra/dist/tarantoolctl connect 3301
```

| Commit                      | Result |
| --------------------------- | ------ |
| 2.10.0-beta2-0-g9d80b68b9   | good   |
| 2.10.0-beta2-1-g0badb4130   | bad    |
| 2.10.0-beta2-2-g8b1aef56a   | bad    |
| 2.10.0-beta2-5-gdc19be406   | bad    |
| 2.10.0-beta2-11-ge493b777d  | bad    |
| 2.10.0-beta2-23-gc9d186082  | bad    |
| 2.10.0-beta2-46-gefee67ca9  | bad    |
| 2.10.0-beta2-92-gee6de025f  | bad    |
| 2.10.0-beta2-184-g2c0aedca0 | bad    |
| 2.10.0-rc1-0-g7ed15e62b     | bad    |

So the problem was introduced by this commit: 0badb4130deac46d6651cdcecb6c6d7ee401a96a. So net.box itself is likely affected and maybe it worth to return the old behaviour back in it. Anyway, I didn't dig into details much and I know who knows all the details: the author of the commit. Please, ask a review from him.

Comment (locker) [reply src/box/lua/console.lua:823]: AFAIK `localhost` isn't quite the same as no host, see [`getaddrinfo(3)`](https://man7.org/linux/man-pages/man3/getaddrinfo.3.html). I think we should fix the net.box to accept nil for host:

```diff
diff --git a/src/box/lua/net_box.lua b/src/box/lua/net_box.lua
index 570108fa5380..b9c4fc2c9324 100644
--- a/src/box/lua/net_box.lua
+++ b/src/box/lua/net_box.lua
@@ -138,8 +138,8 @@ local function parse_connect_params(host_or_uri, ...) -- self? host_or_uri port?
                         type(host_or_uri) == 'number' or
                         type(host_or_uri) == 'table') then
         uri = host_or_uri
-    elseif type(host_or_uri) == 'string' and (type(port) == 'string' or
-                                              type(port) == 'number') then
+    elseif (type(host_or_uri) == 'string' or host_or_uri == nil) and
+           (type(port) == 'string' or type(port) == 'number') then
         uri = urilib.format({host = host_or_uri, service = tostring(port)})
     else
         box.error(E_PROC_LUA,
```

This won't work though, because surprisingly `uri.format` doesn't work if there's no host given, just a service:

```
tarantool> uri.parse(uri.format(uri.parse(3301)))
---
- null
- 'Incorrect URI: expected host:service or /unix.socket'
...
```

This should probably be fixed, as well.

**Review by locker — CHANGES_REQUESTED** [review]
> NO_TEST=bugfix

Lame excuse. Please add a test.

File: changelogs/unreleased/igrishnov/gh-7479-bug-fix-default-uri-tarantoolctl-connect.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+## bugfix/tarantoolctl
+
+* Made host=localhost by default for tarantoolctl connect.
+  Now, when specifying only the port, the connection is made
+  to localhost. This bugfix required to supplement the
+  implementation of the `src/lib/uri` parser.
+  Before this fix a call `uri.parse(uri.format(uri.parse(3301)))`
+  returned an error of 'Incorrect URI'.
+  Now this call return correct `service: '3301'` (gh-7479).
```
Comment (locker) [inline changelogs/unreleased/igrishnov/gh-7479-bug-fix-default-uri-tarantoolctl-connect.md:9]: Nit: No need to go deep into details here - it's a changelog, not a documentation or ticket, it's supposed to be brief. You can leave it as is though - up to you.

File: src/box/lua/net_box.lua [hunk 4]
```diff
@@ -138,8 +138,8 @@ local function parse_connect_params(host_or_uri, ...) -- self? host_or_uri port?
                         type(host_or_uri) == 'number' or
                         type(host_or_uri) == 'table') then
         uri = host_or_uri
-    elseif type(host_or_uri) == 'string' and (type(port) == 'string' or
-                                              type(port) == 'number') then
+    elseif (type(host_or_uri) == 'string' or host_or_uri == nil) and
+            (type(port) == 'string' or type(port) == 'number') then
```
Comment (locker) [inline src/box/lua/net_box.lua:142]: I don't see a test that would cover this code change. Please add a test for the net.box module.

It'd also be nice to have a test for `console.connect` so that we don't occasionally re-introduce this degradation in future.

**Review by locker — CHANGES_REQUESTED** [review]

File: changelogs/unreleased/igrishnov/gh-7479-bug-fix-default-uri-tarantoolctl-connect.md (hunk 3 above)
Comment (GRISHNOV) [reply changelogs/unreleased/igrishnov/gh-7479-bug-fix-default-uri-tarantoolctl-connect.md]: Done, simplified

File: src/box/lua/net_box.lua (hunk 4 above)
Comment (GRISHNOV) [reply src/box/lua/net_box.lua:142]: Added tests for `src/lib/uri` with case `uri.format(uri.parse(3301))`.
Also added tests for `tarantoolctl connect` and `net.box.connect` with only port indication.

File: changelogs/unreleased/igrishnov/gh-7479-bug-fix-uri-lib.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+## bugfix/uri
+
+* Supplement the implementation of the `src/lib/uri` parser.
+  Before this fix a call `uri.parse(uri.format(uri.parse(3301)))`
+  returned an error of 'Incorrect URI'.
+  Now this call return correct `service: '3301'` (gh-7479).
```
Comment (locker) [inline changelogs/unreleased/igrishnov/gh-7479-bug-fix-uri-lib.md:6]: First, changelogs should state changes in Past Simple ("Supplemet**ed** the implementation ...").

Second, we should emphasize what the issue was about. I'd rephrase to something like this:

- Fixed a bug in the URI parser, because of which tarantoolctl failed to connect when the host name was skipped (gh-7479).

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        }
+        -- Connection will be established in a separate process.
+        local ph_tarantoolctl = require('popen').new(cmd_tarantoolctl, {
+            stdin  = 'devnull',
+            stdout = 'devnull',
+        })
+        os.execute("sleep " .. tonumber(1))
+        ph_test_instance:close()
+        ph_tarantoolctl:close()
+    ]==]
```
Comment (locker) [inline test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua:58]: Better put the source in a separate file instead of writing it to a file from a string on the fly. Anyway, I don't understand why you need a separate source file. Can't you create a [server](https://github.com/tarantool/tarantool/blob/master/test/luatest_helpers/server.lua) and start a console in it?

```lua
require('console').listen(0)
```

BTW this function returns the port, the console listens to.

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+    local file_remote, err = io.open('/tmp/remote.lua', 'w')
```
Comment (locker) [inline test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua:60]: All temporary files should be written to the test tmp directory, not to the global tmp dir.

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        -- Get test instance init message that contains port information.
+        local test_instance_init_msg = ph_test_instance:read({stderr = true}):rstrip()
+        -- Get substring like *bound to [::]:49803* in instance init message.
+        local port = string.match(test_instance_init_msg, "[::]:]:%s*(%S+)")
+        if port == nil then
+            -- Get substring like *bound to 0.0.0.0:49803* in instance init message.
+            port = string.match(test_instance_init_msg, "0.0.0.0:%s*(%S+)")
+        end
+        file:write(port)
+        os.execute("sleep " .. tonumber(1))
```
Comment (locker) [inline test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua:42]: We have [fio](https://www.tarantool.io/en/doc/latest/reference/reference_lua/fio/) and [fiber](https://www.tarantool.io/en/doc/latest/reference/reference_lua/fiber/) modules, which make life much easier. For example, you could use `fiber.sleep` here.

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+    if port == nil then
+        error('probably a problem with parsing the port value')
+    end
+    t.assert_equals(result, 'connected to localhost:' .. port)
+end
+
+-- Checks whether it is possible to specify only port for net.box.connect.
+g.test_net_box_connect_with_only_port_indication = function()
+    local ok, msg = pcall(net_box.connect, nil, 3301)
```
Comment (locker) [inline test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua:87]: Please don't connect to 3301 in a test - the port may actually be open on the test machine, in which case the test would hang.

**Review by locker — CHANGES_REQUESTED** [review]

File: changelogs/unreleased/igrishnov/gh-7479-bug-fix-uri-lib.md (hunk 5 above)
Comment (GRISHNOV) [reply changelogs/unreleased/igrishnov/gh-7479-bug-fix-uri-lib.md]: Done

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua (hunk 6 above)
Comment (GRISHNOV) [reply test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua]: Done.  Thanks for pointing to `console.listen(0)`, it really simplified the solution

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua (hunk 7 above)
Comment (GRISHNOV) [reply test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua]: Done. Now tmp dir is not used

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua (hunk 8 above)
Comment (GRISHNOV) [reply test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua]: Done. Now sleep is not used

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua (hunk 9 above)
Comment (GRISHNOV) [reply test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua]: Done

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+local popen   = require('popen')
+local net_box = require('net.box')
+local server  = require('test.luatest_helpers.server')
+
+-- Create test instance.
+g.before_all = function()
+    g.server = server:new({alias = 'test-gh-7479'})
+    g.server:start()
+    -- Get port for connection testing.
+    g.server_port = g.server:eval("return require('console').listen(0):name()['port']")
```
Comment (locker) [inline test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua:12]: Please use `g.server:exec()` instead of eval.

Also, please keep all lines within 80 characters.

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+end
+
+-- Stop test instance.
+g.after_all = function()
+    g.server:stop()
+end
+
+-- Checks whether it is possible to specify only port for tarantoolctl connect.
+g.test_tarantoolctl_connect = function()
+    local tarantoolctl_location = io.popen('which tarantoolctl'):read()
```
Comment (locker) [inline test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua:22]: We want to test tarantoolctl from the build:

```Lua
local tarantoolctl_location = fio.pathjoin(os.getenv('BUILDDIR'),
                                           'extra', 'dist', 'tarantoolctl')
```

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+    t.assert_equals(result, 'connected to localhost:' .. tostring(g.server_port))
+end
+
+-- Checks whether it is possible to specify only port for net.box.connect.
+g.test_net_box_connect_with_only_port_indication = function()
+    -- Indicate only port, without localhost:* before the port value.
+    local ok, msg = pcall(net_box.connect, nil, g.server_port)
+    t.assert_equals(ok, true)
+    t.assert_is_not(tostring(msg), 'usage: connect(uri[, opts] | host, port[, opts])')
```
Comment (locker) [inline test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua:48]: No need to check the error message - the error code is enough. (We may change the error message in future, in which case the check would become pointless.)

**Review by locker — CHANGES_REQUESTED** [review]

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua (hunk 10 above)
Comment (GRISHNOV) [reply test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua]: Done

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua (hunk 12 above)
Comment (GRISHNOV) [reply test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua]: Done

File: test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua (hunk 11 above)
Comment (GRISHNOV) [reply test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua]: Done, but it began to work only with the `tarantoolctl.in` instead of `tarantoolctl`. AFAIU, it is ok?

Comment (locker) [reply test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua]: No, it isn't. `tarantoolctl.in` is a source file - it shouldn't be executed. Here's how we find `tarantoolctl` in another test:

https://github.com/tarantool/tarantool/blob/ac5f303d2b65dde0d1fa889ad716b4c7f154846c/test/app-tap/tarantoolctl.test.lua#L16-L31

Comment (GRISHNOV) [reply test/app-luatest/gh_7479_bug_fix_uri_lib_test.lua]: Done. Thanks! It works with the option `--builddir`

**Review by locker — APPROVED** [review]

**Comment by locker** [comment]
Cherry-picked to 2.10.
